documentation/src/main/asciidoc/userguide/chapters/query/hql/QueryLanguage.adoc

```diff
@@ -147,77 +190,105 @@ include::{sourcedir}/../batch/BatchTest.java[tags=batch-bulk-hql-update-version-
 ====
 
-The BNF for an HQL `INSERT` statement is:
+The BNF for an `insert` statement is:
 
 [[hql-insert-bnf-example]]
 ====
-[source, SQL, indent=0]
+[source, antlrv4, indent=0]
 ----
 include::{extrasdir}/statement_insert_bnf.txt[]
 ----
 ====
 
-The `attribute_list` is analogous to the `column specification` in the SQL `INSERT` statement.
-For entities involved in mapped inheritance, only attributes directly defined on the named entity can be used in the `attribute_list`.
-Superclass properties are not allowed and subclass properties do not make sense.
-In other words, `INSERT` statements are inherently non-polymorphic.
-
-`select_statement` can be any valid HQL select query, with the caveat that the return types must match the types expected by the insert.
-Currently, this is checked during query compilation rather than allowing the check to delegate to the database.
-This may cause problems between Hibernate Types which are _equivalent_ as opposed to __equal__.
-For example, this might lead to issues with mismatches between an attribute mapped as a `org.hibernate.type.StandardBasicTypes.DATE` and an attribute defined as a `org.hibernate.type.StandardBasicTypes.TIMESTAMP`,
-even though the database might not make a distinction or might be able to handle the conversion.
-
-For the id attribute, the insert statement gives you two options.
-You can either explicitly specify the id property in the `attribute_list`, in which case its value is taken from the corresponding select expression, or omit it from the `attribute_list` in which case a generated value is used.
-This latter option is only available when using id generators that operate "in the database"; attempting to use this option with any "in memory" type generators will cause an exception during parsing.
-
-For optimistic locking attributes, the insert statement again gives you two options.
-You can either specify the attribute in the `attribute_list` in which case its value is taken from the corresponding select expressions or omit it from the `attribute_list` in which case the `seed value` defined by the corresponding `org.hibernate.type.VersionType` is used.
+For example:
 
 [[hql-insert-example]]
-.INSERT query statements
 ====
+[source, SQL, indent=0]
+----
+include::{sourcedir}/HQLTest.java[tags=hql-insert-example]
+----
+
 [source, SQL, indent=0]
 ----
 include::{sourcedir}/../batch/BatchTest.java[tags=batch-bulk-hql-insert-example]
 ----
 ====
 
+An `insert` statement must be executed by calling `Query#executeUpdate()`.
+
+[IMPORTANT]
+====
+`insert` statements are inherently _not_ polymorphic!
+The `targetFields` list is of fixed length, whereas each subclass of an entity class might declare additional fields.
+For an entity involved in a mapped inheritance hierarchy, only attributes declared directly by the named entity may be used in the list of target fields.
+Superclass attributes are not allowed, and subclass attributes do not make sense.
```

Comment:
>Superclass attributes are not allowed

Not sure this is still the case. I guess you just moved this to here from somewhere? Assigning superclass attributes is definitely possible in the insert statement just as well as assigning fields with columns contained in secondary tables. This is new and I just recently implemented this.

Reply:
OK, good, it seemed strange to me that this would be the case.

So can I completely remove that passage or are there still limitations the user needs to know about?

Reply:
I don't think that there are any limitations anymore.

Reply:
I presume that it's still not possible to assign to _subclass_ fields.

Reply:
https://github.com/hibernate/hibernate-orm/pull/4558

Reply:
> I presume that it's still not possible to assign to subclass fields.

AFAIU it doesn't make sense to do that 😄